devito/types/basic.py

```diff
@@ -1170,7 +1176,8 @@ def bound_symbols(self):
     @cached_property
     def indexed(self):
         """The wrapped IndexedData object."""
-        return IndexedData(self.name, shape=self._shape, function=self.function)
+        return IndexedData(self.name, shape=self._shape, function=self.function,
+                           dtype=self.dtype)
```

Comment:
this should be carried by `self.function` already... how can it be different?

Reply:
It is, can leave it out since it will use `function.dtype` if not provided so doesn't really matter